kong/clustering/wrpc_control_plane.lua

```diff
@@ -0,0 +1,690 @@
+local _M = {}
+
+
+local semaphore = require("ngx.semaphore")
+local ws_server = require("resty.websocket.server")
+local ssl = require("ngx.ssl")
+local ocsp = require("ngx.ocsp")
+local http = require("resty.http")
+local cjson = require("cjson.safe")
+local declarative = require("kong.db.declarative")
+local constants = require("kong.constants")
+local openssl_x509 = require("resty.openssl.x509")
+local wrpc = require("kong.tools.wrpc")
+local string = string
+local setmetatable = setmetatable
+local type = type
+local pcall = pcall
+local pairs = pairs
+local ipairs = ipairs
+local tonumber = tonumber
+local tostring = tostring
+local ngx = ngx
+local ngx_log = ngx.log
+local cjson_encode = cjson.encode
+local kong = kong
+local ngx_exit = ngx.exit
+local exiting = ngx.worker.exiting
+local ngx_time = ngx.time
+local ngx_var = ngx.var
+local table_insert = table.insert
+local table_concat = table.concat
+
+local kong_dict = ngx.shared.kong
+local KONG_VERSION = kong.version
+local ngx_DEBUG = ngx.DEBUG
+local ngx_INFO = ngx.INFO
+local ngx_NOTICE = ngx.NOTICE
+local ngx_WARN = ngx.WARN
+local ngx_ERR = ngx.ERR
+local ngx_CLOSE = ngx.HTTP_CLOSE
+local MAX_PAYLOAD = constants.CLUSTERING_MAX_PAYLOAD
+local WS_OPTS = {
+  timeout = constants.CLUSTERING_TIMEOUT,
+  max_payload_len = MAX_PAYLOAD,
+}
+local OCSP_TIMEOUT = constants.CLUSTERING_OCSP_TIMEOUT
+local CLUSTERING_SYNC_STATUS = constants.CLUSTERING_SYNC_STATUS
+local MAJOR_MINOR_PATTERN = "^(%d+)%.(%d+)%.%d+"
+local _log_prefix = "[clustering] "
+
+local wrpc_config_service
+
+local function get_config_service(self)
+  if not wrpc_config_service then
+    wrpc_config_service = wrpc.new_service()
+    wrpc_config_service:add("kong.services.config.v1.config")
```

Comment:
a new protocol version would require new code.  i think it's best that each version of the code hardcodes the appropriate version of the .proto; in fact, .proto files are part of the code, even if loaded dynamically instead of generating static code as on Go